system/src/Grav/Common/Page/Medium/ImageMedium.php

```diff
@@ -232,31 +218,36 @@ public function srcset($reset = true)
      * @return $this
      */
     public function derivatives($min_width, $max_width, $step = 200) {
-      $width = $min_width;
-
-      // Do not upscale images.
-      if ($max_width > $this->get('width')) {
-        $max_width = $this->get('width');
-      }
-
-      while ($width <= $max_width) {
-        $ratio = $width / $this->get('width');
-        $derivative = MediumFactory::scaledFromMedium($this, 1, $ratio);
-        if (is_array($derivative)) {
-          $this->addDerivative($derivative['file']);
+        $width = $min_width;
+
+        // Do not upscale images.
+        if ($max_width > $this->get('width')) {
+            $max_width = $this->get('width');
         }
-        $width += $step;
-      }
-      return $this;
-    }
 
-    /**
-     * Add a derivative
-     *
-     * @param  ImageMedium $image
-     */
-    public function addDerivative(ImageMedium $image) {
-      $this->derivatives[$image->url()] = $image->get('width');
+        // Clear any alternatives that have already been implied by the image's
+        // filename (eg. ones that had "@2x" or similar appended to their
+        // filenames)
+        $this->clearAlternatives();
+
+        while ($width <= $max_width) {
+            $ratio = $width / $this->get('width');
+            $width += $step;
+
+            if ($ratio == 1) {
+                continue;
+            }
+
+            $derivative = MediumFactory::fromFile($this->get('filepath'));
```

Comment:
After some pretty basic profiling of a very image heavy page, I'm a bit concerned with this method. Reading the image from disk every time seems to have a significant impact on performance, almost on par with the image processing itself.

This PR saves an additional disk write, which *could* be beneficial for performance, I guess. Since reading/writing is slow, avoiding touching disk is of course good, but with `MediumFactory::scaledFromMedium`, the file that we're reading might be getting incrementally smaller, which could also be beneficial for performance if the original image is very large.